tsfc/ufl_utils.py

```diff
@@ -79,19 +83,27 @@ def one_times(measure):
     return integrand, degree
 
 
-def preprocess_expression(expression):
+def preprocess_expression(expression, complex_mode=False):
     """Imitates the compute_form_data processing pipeline.
 
     Useful, for example, to preprocess non-scalar expressions, which
     are not and cannot be forms.
     """
+    if complex_mode:
+        expression = do_comparison_check(expression)
+    if not complex_mode:
+        expression = remove_complex_nodes(expression)
     expression = apply_algebra_lowering(expression)
     expression = apply_derivatives(expression)
     expression = apply_function_pullbacks(expression)
     expression = apply_geometry_lowering(expression, preserve_geometry_types)
     expression = apply_derivatives(expression)
     expression = apply_geometry_lowering(expression, preserve_geometry_types)
     expression = apply_derivatives(expression)
+    if not complex_mode:
+        expression = remove_complex_nodes(expression)
```

Comment:
Why are we removing complex nodes twice?

Reply:
I think it's because intermediate steps can add new complex nodes.

Reply:
What about doing just the second removal then?

Reply:
I think Nic was concerned that complex nodes might disable simplifications that might otherwise occur.